django/test/utils.py

```diff
@@ -170,71 +172,140 @@ def teardown_test_environment():
+        self.old_names = []
+        self.main_db_created = threading.Event()
+
+    def setup(self):
+        """Main method to set up test databases."""
+        self.test_databases, self.mirrored_aliases = get_unique_databases_and_mirrors(
+            self.aliases
+        )
+
+        # Process each unique database
+        for db_sig, (db_name, aliases) in self.test_databases.items():
+            # Extract the first alias outside the loop
+            first_alias = aliases[0]
+            remaining_aliases = aliases[1:]
 
-    # Configure the test mirrors.
-    for alias, mirror_alias in mirrored_aliases.items():
-        connections[alias].creation.set_as_test_mirror(
-            connections[mirror_alias].settings_dict
+            # Create the test database for the first alias
+            self.create_test_database(first_alias, db_name)
+
+            # Clone the test database if parallel testing is enabled
+            if self.parallel > 1:
+                self.clone_databases_in_parallel(first_alias)
+
+            # Configure remaining aliases as mirrors
+            for alias in remaining_aliases:
+                self.configure_as_mirror(alias, first_alias)
+
+        # Configure any additional test mirrors
+        self.configure_test_mirrors()
+
+        # Enable debug SQL logging if required
+        if self.debug_sql:
+            self.enable_debug_sql()
+
+        return self.old_names
+
+    def create_test_database(self, alias, db_name):
+        """Create the test database for the given alias."""
+        connection = connections[alias]
+        self.old_names.append((connection, db_name, True))
+
+        serialize_alias = (
+            self.serialized_aliases is None or alias in self.serialized_aliases
         )
 
-    if debug_sql:
+        with self.time_keeper.timed(f"  Creating '{alias}'"):
+            connection.creation.create_test_db(
+                verbosity=self.verbosity,
+                autoclobber=not self.interactive,
+                keepdb=self.keepdb,
+                serialize=serialize_alias,
+            )
+
+        self.main_db_created.set()
+
+    def clone_databases_sequentially(self, alias):
+        """Clone the test database sequentially for parallel test execution."""
+        for index in range(self.parallel):
+            self.clone_test_database(alias, index)
+
+    def clone_databases_in_parallel(self, alias):
+        """Clone the test database in parallel threads for parallel test execution."""
+        threads = []
+        for index in range(self.parallel):
+            t = threading.Thread(
+                target=self.clone_test_database,
+                args=(alias, index),
+            )
+            t.start()
+            threads.append(t)
+
+        for t in threads:
+            t.join()
+
+    def clone_test_database(self, alias, index):
+        """Clone the test database for parallel execution."""
+
+        self.main_db_created.wait()
```

Comment:
Question: It looks like this will only ever be run once the main DB exists, since the threads are only started. Could this event be removed?

Reply:
@RealOrangeOne My idea is that I'm waiting for the db to be there, then clone, as all the threads are dependant of the main db, does this answer or maybe can you elaborate?

Reply:
From my reading, you're not starting the threads until after the DB is created, so the wait event isn't necessary.

Reply:
Thank you